packages/flutter/lib/src/material/predictive_back_page_transitions_builder.dart

```diff
@@ -0,0 +1,299 @@
+  }
+
+  /// The back event when the gesture first started.
+  PredictiveBackEvent? get startBackEvent => _startBackEvent;
+  set startBackEvent(PredictiveBackEvent? startBackEvent) {
+    if (_startBackEvent != startBackEvent && mounted) {
+      setState(() {
+        _startBackEvent = startBackEvent;
+      });
+    }
+  }
+
+  /// The most recent back event during the gesture.
+  PredictiveBackEvent? _currentBackEvent;
+  PredictiveBackEvent? get currentBackEvent => _currentBackEvent;
+  set currentBackEvent(PredictiveBackEvent? currentBackEvent) {
+    if (_currentBackEvent != currentBackEvent && mounted) {
+      setState(() {
+        _currentBackEvent = currentBackEvent;
+      });
+    }
+  }
+
+  // Begin WidgetsBinding.
+
+  @override
+  bool handleStartBackGesture(PredictiveBackEvent backEvent) {
+    _gestureInProgress = !backEvent.isButtonEvent && _isEnabled;
+    if (!_gestureInProgress) {
+      return false;
+    }
+
+    widget.predictiveBackRoute.handleStartBackGesture(progress: 1 - backEvent.progress);
+    startBackEvent = currentBackEvent = backEvent;
+    return true;
+  }
+
+  @override
+  bool handleUpdateBackGestureProgress(PredictiveBackEvent backEvent) {
+    if (!_gestureInProgress) {
+      return false;
+    }
+
+    widget.predictiveBackRoute.handleUpdateBackGestureProgress(progress: 1 - backEvent.progress);
+    currentBackEvent = backEvent;
+    return true;
+  }
+
+  @override
+  bool handleCancelBackGesture() {
+    if (!_gestureInProgress) {
+      return false;
+    }
+
+    widget.predictiveBackRoute.handleDragEnd(animateForward: true);
+    _gestureInProgress = false;
+    startBackEvent = currentBackEvent = null;
+    return true;
+  }
+
+  @override
+  bool handleCommitBackGesture() {
+    if (!_gestureInProgress) {
+      return false;
+    }
+
+    widget.predictiveBackRoute.handleDragEnd(animateForward: false);
+    _gestureInProgress = false;
+    startBackEvent = currentBackEvent = null;
+    return true;
+  }
+
+  // End WidgetsBinding.
+
+  @override
+  void initState() {
+    super.initState();
+    WidgetsBinding.instance.addObserver(this);
+  }
+
+  @override
+  void dispose() {
+    WidgetsBinding.instance.removeObserver(this);
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return widget.builder(context);
+  }
+}
+
+/// Android's predictive back page transition.
+class _PredictiveBackPageTransition extends StatelessWidget {
+  const _PredictiveBackPageTransition({
+    required this.animation,
+    required this.secondaryAnimation,
+    required this.getIsCurrent,
+    required this.child,
+  });
+
+  final Animation<double> animation;
+  final Animation<double> secondaryAnimation;
+  final ValueGetter<bool> getIsCurrent;
+  final Widget child;
+
+  Widget _secondaryAnimatedBuilder(BuildContext context, Widget? child) {
+    final Size size = MediaQuery.sizeOf(context);
+    final double screenWidth = size.width;
```

Comment:
I think it still makes sense to use the full screen width if you wanted to do this on a non-full-screen route, because it's based on the distance of a drag gesture across the whole screen.

Also just to double check, by default something like showDialog won't use the predictive back gesture, though you could set something like that up if you wanted to.